radix-engine/src/engine/call_frame.rs

```diff
@@ -52,20 +52,37 @@ pub struct CallFrame<
     /// Owned Values
     buckets: HashMap<BucketId, Bucket>,
     proofs: HashMap<ProofId, Proof>,
-    owned_values: ComponentObjects,
+    owned_values: HashMap<StoredValueId, StoredValue>,
 
     /// Referenced values
     worktop: Option<Worktop>,
     auth_zone: Option<AuthZone>,
     component_state: Option<&'p mut ComponentState>,
-    ref_values: HashSet<StoredValueId>,
+    refed_values: HashMap<StoredValueId, ValueRefType>,
 
     /// Caller's auth zone
     caller_auth_zone: Option<&'p AuthZone>,
 
     phantom: PhantomData<I>,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ValueType {
+    Owned,
+    Ref(ValueRefType),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ValueRefType {
+    Uncommitted {
+        root: KeyValueStoreId,
+        ancestors: Vec<KeyValueStoreId>,
```

Comment:
Minor: can be merged into `path: Vec<KeyValueStoreId>`?

Reply:
I had that originally but this way encodes the invariant that there must always be a root value at the compile time level (rather than requiring use of `unwrap()` later)

Reply:
I see. Let's leave it for now then